File: leetcode/algorithms/482_license_key_formatting/main.dart
class LicenseKeyFormatting {
  String licenseKeyFormatting(String s, int k) {
    String result = "";
    List<String> sArray = s.replaceAll("-", "").split("");

    for (int i = 1; i <= sArray.length; i++) {
      if (i < sArray.length && (i % k) == 0) {
        result = "-${sArray[sArray.length - i].toUpperCase()}$result";
      } else {
        result = "${sArray[sArray.length - i].toUpperCase()}$result";
      }
    }

    return result;
  }


  // Solution
  String solution(String s, int k) {
    final _dash = '-'.codeUnits[0];

    var result = <int>[];
    var currentGroupLength = 0;

    for (var char in s.codeUnits.reversed) {
      if (char == _dash) {
        continue;
      }
      result.add(char);
      currentGroupLength++;
      if (currentGroupLength == k) {
        result.add(_dash);
        currentGroupLength = 0;
      }
    }

    if (!result.isEmpty && result.last == _dash) {
      result.removeLast();
    }

    return String.fromCharCodes(result.reversed).toUpperCase();
  }
}

File: leetcode/algorithms/482_license_key_formatting/main.java
class LicenseKeyFormatting {
    public String licenseKeyFormatting(String s, int k) {
        s = s.replaceAll("-", "").toUpperCase();
        String result = "";

        for (int i = s.length(); i >= 0; i -= k) {
            if (i > k) {
                result = "-" + s.substring(i - k, i) + result;
            } else {
                result = s.substring(0, i) + result;
            }
        }

        return result;
    }


    // Solution
    public String solution(String s, int k) {
        StringBuilder sb = new StringBuilder();
        for (int i = s.length() - 1; i >= 0; i--) {
            if (s.charAt(i) != '-') {
                sb.append(sb.length() % (k + 1) == k ? '-' : "").append(s.charAt(i));
            }
        }
        return sb.reverse().toString().toUpperCase();
    }
}
